packages/network-controller/src/NetworkController.ts

```diff
@@ -136,6 +211,110 @@ function isInfuraProviderType(type: string): type is InfuraNetworkType {
   return Object.keys(InfuraNetworkType).includes(type);
 }
 
+/**
+ * Builds an identifier for an Infura network client for lookup purposes.
+ *
+ * @param infuraNetworkOrProviderConfig - The name of an Infura network or a
+ * provider config.
+ * @returns The built identifier.
+ */
+function buildInfuraNetworkClientId(
+  infuraNetworkOrProviderConfig:
+    | InfuraNetworkType
+    | (ProviderConfig & { type: InfuraNetworkType }),
+): BuiltInNetworkClientId {
+  if (typeof infuraNetworkOrProviderConfig === 'string') {
+    return `infura||${infuraNetworkOrProviderConfig}` as const;
+  }
+  return `infura||${infuraNetworkOrProviderConfig.type}` as const;
+}
+
+/**
+ * Builds an identifier for a custom network client for lookup purposes.
+ *
+ * @param args - This function can be called two ways:
+ * 1. The ID of a network configuration.
+ * 2. A provider config and a set of network configurations.
+ * @returns The built identifier.
+ */
+function buildCustomNetworkClientId(
+  ...args:
+    | [NetworkConfigurationId]
+    | [
+        ProviderConfig & { type: typeof NetworkType.rpc; rpcUrl: string },
+        NetworkConfigurations,
+      ]
+): CustomNetworkClientId {
+  if (args.length === 1) {
+    return `custom||${args[0]}` as const;
+  }
+  const [{ id, rpcUrl, chainId }, networkConfigurations] = args;
+  if (id === undefined) {
+    const matchingNetworkConfiguration = Object.values(
+      networkConfigurations,
+    ).find((networkConfiguration) => {
+      return networkConfiguration.rpcUrl === rpcUrl.toLowerCase();
+    });
+    if (matchingNetworkConfiguration) {
+      return `custom||${matchingNetworkConfiguration.id}`;
+    }
+    return `custom||${chainId}||${rpcUrl.toLowerCase()}` as const;
```

Comment:
Nit: would just the RPC URL suffice in this case? At the moment it's not possible to use one URL for multiple chains.

Reply:
Yup! Fixed in [9e67774](https://github.com/MetaMask/core/pull/1439/commits/9e67774278ae3433014f286fb8631e3a4c7eddec).